dynamic_instrumentation/src/instrument_memory.rs

```diff
@@ -437,32 +459,47 @@ impl<'a, 'tcx: 'a> Visitor<'tcx> for FunctionInstrumenter<'a, 'tcx> {
-
-                    error: internal compiler error: compiler/rustc_borrowck/src/borrow_set.rs:252:17:
-                    found two uses for 2-phase borrow temporary _9: bb45[0] and bb46[1]
-                    --> ../analysis/test/src/pointers.rs:162:16
-                        |
-                    162 |     for arg in ::std::env::args() {
-                        |                ^^^^^^^^^^^^^^^^^^
-                */
-
-                // location.statement_index += 1;
-                // if let Some(p) = rv_place(value) {
-                //     self.add_instrumentation(
-                //         location,
-                //         ref_copy_fn,
-                //         vec![Operand::Copy(dest)],
-                //         false,
-                //         false,
-                //         EventMetadata {
-                //             source: Some(to_mir_place(&p)),
-                //             destination: Some(to_mir_place(&dest)),
-                //         },
-                //     );
-                // }
+            } else if value.ty(&self.body.local_decls, self.tcx).is_region_ptr() { // Rhs is a ref
+                /// Used to strip initital deref from projection sequences
+                fn pop_one_projection<'tcx>(p: &Place<'tcx>, tcx: TyCtxt<'tcx>) -> Option<Place<'tcx>> {
+                    if p.projection.is_empty() {
+                        return None;
+                    }
+                    Some(Place {
+                        local: p.local,
+                        projection: tcx.intern_place_elems(&p.projection[..p.projection.len() - 1]),
+                    })
+                }
+                // For mutable borrows (let _n = &mut place), create a parallel binding that takes
+                // a raw pointer to the same place, without involving _n
+                if let Rvalue::Ref(_, BorrowKind::Mut {..}, p) = value {
+                    // We do not increment the statement index because we want to take a raw ptr
+                    // prior to the existing stmt that takes a mutable pointer
+                    //location.statement_index += 1;
+
+                    // Remove outer deref if present, so we turn `&mut *x` into `addr_of!(x)` rather
+                    // than `addr_of!(*x)`
+                    let arg = match p.iter_projections().last() {
+                        Some((_, ProjectionElem::Deref)) => {
+                            let sans_proj = pop_one_projection(p, self.tcx.clone())
+                                .expect("expected but did not find deref projection");
+                            Operand::Copy(sans_proj)
+                        },
+                        _ => Operand::Copy(p.clone()),
+                    };
+                    self.add_instrumentation_point(
+                        location,
+                        ref_copy_fn,
+                        vec![InstrumentationOperand::Place(arg)],
```

Comment:
should this be InstrumentationOperand::Ref

Reply:
Rebased.